### PR TITLE
kvdb-rocksdb: pass ReadOptions to iterators

### PR DESCRIPTION
**Pull request by ordian**
Passing `prefix_same_as_start` has no effect, since we're not using "Prefix Seek" mode, see https://github.com/facebook/rocksdb/blob/9befbe9b406574e8d8b687799701bd8cebf5163b/include/rocksdb/options.h#L1246-L1251 and https://github.com/facebook/rocksdb/wiki/Prefix-Seek-API-Changes.

### CODE REVIEW

**Comment by ordian** [comment]
> audit code using the prefix search and figure out if it really needs to; if, as I suspect, we're not using this

we are: [1](https://github.com/paritytech/parity-ethereum/blob/f6c3d4c6951ecf2070bdf7e6a5c6a24edb1a43d6/ethcore/blockchain/src/blockchain.rs#L1006-L1015), [2](https://github.com/paritytech/parity-ethereum/blob/f6c3d4c6951ecf2070bdf7e6a5c6a24edb1a43d6/util/journaldb/src/overlayrecentdb.rs#L281-L287).

> In order to use prefix search users must create a prefix extractor (rocksdb::SliceTransform::create_fixed_prefix(…)) when opening the DB and then pass it along when instantiating an iterator, is that correct?

 yes
> The test_prefix_iterator() test in rust-rocksdb uses a different type though: Options (i.e. ffi::rocksdb_options_t) – do both work?

`set_prefix_extractor` is set on `Options` (this is a global option, which affects the db layout, db is partitioned based on prefix),
`set_prefix_same_as_start` is called on `ReadOptions` when creating an Iterator
https://docs.rs/rocksdb/0.13.0/src/rocksdb/db.rs.html#1153

**Comment by ordian** [comment]
> (with more docs/examples illustrating usage)

what do you mean, we haven't changed to public API here

**Comment by dvdplm** [comment]
> what do you mean, we haven't changed the public API here

True, but the docs for the existing API are lacking and do not mention how to set up the prefix search properly. I think we want an example [here](https://github.com/paritytech/parity-common/pull/277/files#diff-3a3dcd8bf87a5b5c4097e2d12a129e17R80-R81), docs [here](https://github.com/paritytech/parity-common/pull/277/files#diff-3a3dcd8bf87a5b5c4097e2d12a129e17R93) and probably module level docs discussing the quirks of prefix search too?

**Comment by dvdplm** [comment]
> > audit code using the prefix search and figure out if it really needs to; if, as I suspect, we're not using this
> 
> we are: [1](https://github.com/paritytech/parity-ethereum/blob/f6c3d4c6951ecf2070bdf7e6a5c6a24edb1a43d6/ethcore/blockchain/src/blockchain.rs#L1006-L1015), [2](https://github.com/paritytech/parity-ethereum/blob/f6c3d4c6951ecf2070bdf7e6a5c6a24edb1a43d6/util/journaldb/src/overlayrecentdb.rs#L281-L287).

I see. So that code needs to be updated after this PR then? And does it even work right now?

**Comment by ordian** [comment]
> So that code needs to be updated after this PR then?

no logic has changed

But I just realised that we changed the logic in #257 https://github.com/paritytech/parity-common/commit/8fb8f13c8084ba8770dcfadca71579eeaac05685#diff-4382bfb7e75959b2b8884268540abfa3R597, previously `iter_from_prefix` returned an iterator over all keys >= `prefix`, now it only returns an iterator of keys starting with the given prefix. I thought it was the expected behavior, but not sure now :/

**Comment by ordian** [comment]
> But I just realised that we changed the logic in #257

After a quick search, in parity-ethereum we use iter_from_prefix only in [epoch_transitions](https://github.com/paritytech/parity-ethereum/blob/f6c3d4c6951ecf2070bdf7e6a5c6a24edb1a43d6/ethcore/blockchain/src/blockchain.rs#L1006-L1015) and then we filter it to start with prefix [here](https://github.com/paritytech/parity-ethereum/blob/f6c3d4c6951ecf2070bdf7e6a5c6a24edb1a43d6/ethcore/blockchain/src/blockchain.rs#L520-L522). The same thing happens in substrate:
[usage](https://github.com/paritytech/substrate/blob/fc6914cdb8c322cea8a4591bbf9566dd23bfd143/client/src/leaves.rs#L83), [filter](https://github.com/paritytech/substrate/blob/fc6914cdb8c322cea8a4591bbf9566dd23bfd143/client/src/leaves.rs#L84). So even though we've changed the logic in #257, it doesn't affect parity-ethereum and substrate. The question is: do we want the new behavior or the old one?

File: kvdb-rocksdb/src/iter.rs [hunk 1]
```diff
@@ -76,22 +76,27 @@ pub trait IterationHandler {
...
+	fn iter_from_prefix(&self, col: Option<u32>, prefix: &[u8], read_opts: &ReadOptions) -> Self::Iterator;
 }
 
 impl<'a, T> ReadGuardedIterator<'a, <&'a T as IterationHandler>::Iterator, T>
 where
 	&'a T: IterationHandler,
 {
-	pub fn new(read_lock: RwLockReadGuard<'a, Option<T>>, col: Option<u32>) -> Self {
-		Self { inner: Self::new_inner(read_lock, |db| db.iter(col)) }
+	pub fn new(read_lock: RwLockReadGuard<'a, Option<T>>, col: Option<u32>, read_opts: &ReadOptions) -> Self {
```
Comment (dvdplm) [inline kvdb-rocksdb/src/iter.rs]: docs

File: kvdb-rocksdb/src/iter.rs [hunk 2]
```diff
@@ -76,22 +76,27 @@ pub trait IterationHandler {
...
+		Self { inner: Self::new_inner(read_lock, |db| db.iter(col, read_opts)) }
 	}
 
-	pub fn new_from_prefix(read_lock: RwLockReadGuard<'a, Option<T>>, col: Option<u32>, prefix: &[u8]) -> Self {
-		Self { inner: Self::new_inner(read_lock, |db| db.iter_from_prefix(col, prefix)) }
+	pub fn new_from_prefix(
```
Comment (dvdplm) [inline kvdb-rocksdb/src/iter.rs]: docs

File: kvdb-rocksdb/src/lib.rs [hunk 3]
```diff
@@ -588,12 +588,14 @@ impl Database {
...
 		let read_lock = self.db.read();
 		let optional = if read_lock.is_some() {
-			let guarded = iter::ReadGuardedIterator::new_from_prefix(read_lock, col, prefix);
+			let guarded = iter::ReadGuardedIterator::new_from_prefix(read_lock, col, prefix, &self.read_opts);
 			Some(interleave_ordered(Vec::new(), guarded))
 		} else {
 			None
 		};
-		// workaround for https://github.com/facebook/rocksdb/issues/2343
+		// We're not using "Prefix Seek" mode, so the iterator will return
```
Comment (dvdplm) [inline kvdb-rocksdb/src/lib.rs]: I'd duplicate&extend this in the module level docs.

File: kvdb-rocksdb/src/iter.rs [hunk 4]
```diff
@@ -76,22 +80,31 @@ pub trait IterationHandler {
...
 }
 
 impl<'a, T> ReadGuardedIterator<'a, <&'a T as IterationHandler>::Iterator, T>
 where
 	&'a T: IterationHandler,
 {
-	pub fn new(read_lock: RwLockReadGuard<'a, Option<T>>, col: Option<u32>) -> Self {
-		Self { inner: Self::new_inner(read_lock, |db| db.iter(col)) }
+	/// Maps `RwLock<RocksDB>` to `RwLock<DBIterator>`, where
+	/// `DBIterator` iterates over all keys.
```
Comment (dvdplm) [inline kvdb-rocksdb/src/iter.rs]:
```suggestion
	/// `DBIterator` iterates over all keys.
	/// In addition to a read lock and a column index, it takes a ref to `ReadOptions` to allow customization of the iterator behaviour, e.g. <insert illuminating example here>.
```

File: kvdb-rocksdb/src/iter.rs (hunk 1 above)
Comment (dvdplm) [reply kvdb-rocksdb/src/iter.rs]: `<insert illuminating example here>.` – needs some editing... ;)

…and there are other methods that now take `ReadOptions` that needs some docs?

**Comment by dvdplm** [comment]
ping @bkchr

File: kvdb-rocksdb/src/iter.rs [hunk 5]
```diff
@@ -78,22 +84,35 @@ pub trait IterationHandler {
 
 	/// Create an `Iterator` over the default DB column or over a `ColumnFamily` if a column number
 	/// is passed.
-	fn iter(&self, col: u32) -> Self::Iterator;
+	/// In addition to a read lock and a column index, it takes a ref to the same `ReadOptions` we
```
Comment (bkchr) [inline kvdb-rocksdb/src/iter.rs]: Not sure why it talks about the read lock here? That is not part of the interface?

Same below.

**Pull request merged**